File: scripts/content-audit.py
#!/usr/bin/env python3
"""scripts/content-audit.py — mechanical content checks.

A flat list of automated pass/fail checks on component bodies. These are lint
guardrails, NOT the quality rubric: the human review scores the six dimensions
D1-D6 into the `rating` (+ `d6` floor) — see docs/REVIEWER_CHECKLIST.md. This
script just catches objective violations.

  CRITICAL (block merge under --strict):
    - desc_too_long            description > 1024 chars (Anthropic spec cap)
    - body_too_long            body over the type cap (skill 600 / agent 350)
    - unparseable_frontmatter

  WARNING (advisory, never blocks):
    - windows_path             Windows backslash paths in body (vendor allowlist)
    - skill_body_over_500      skill body over Anthropic's 500-line guidance
    - readme_count_mismatch    README component-table row count != SKILL.md +
                               agent .md files on disk. Advisory while the
                               backlog is burned down; promote to CRITICAL once
                               all plugin READMEs enumerate their components.

Usage: python3 scripts/content-audit.py [--strict]
"""
import re, os, glob
from collections import defaultdict

# Body-length ceilings keyed on component TYPE (skill vs agent). Anthropic's
# only published length guidance is "keep SKILL.md under 500 lines"; agents have
# no published cap (they run in their own context). These hard caps are generous
# absolute ceilings — "something is structurally wrong if exceeded".
SKILL_BODY_HARDCAP = 600
AGENT_BODY_HARDCAP = 350
SKILL_BODY_ADVISORY = 500  # Anthropic's published SKILL.md guidance

# Allowlist for windows_path — components where citing a literal Windows path is
# the documented canonical install/usage location for a Windows-only tool, not a
# hygiene oversight. Reviewed 2026-05-27; vendor-canonical paths only.
WINDOWS_PATH_ALLOWLIST = {
    "plugins/qa-desktop/skills/appium-windows-driver/SKILL.md",
    "plugins/qa-desktop/skills/winappdriver/SKILL.md",
    "plugins/qa-game/skills/unity-test-framework/SKILL.md",
    "plugins/qa-desktop/skills/desktop-test-strategy-reference/SKILL.md",
}


def parse_frontmatter(path):
    try:
        with open(path, encoding="utf-8") as f:
            c = f.read()
    except Exception:
        return None, ""
    if not c.startswith("---"):
        return None, c
    parts = c.split("---", 2)
    if len(parts) < 3:
        return None, c
    fm_text, body = parts[1], parts[2]
    m = re.search(r'description:\s*"(.*?)"\s*\n[a-z_-]+:', fm_text, re.DOTALL)
    if m:
        desc = m.group(1)
    else:
        m2 = re.search(r'^description:\s*"?(.*?)"?\s*$', fm_text, re.MULTILINE)
        desc = m2.group(1) if m2 else ""
    return {"_description_raw": desc}, body


def find_components():
    paths = []
    for pat in ["plugins/*/skills/*/SKILL.md", "plugins/*/agents/*.md"]:
        paths.extend(p.replace("\\", "/") for p in glob.glob(pat))
    return sorted(paths)


def audit():
    findings = defaultdict(list)
    total = 0
    for p in find_components():
        total += 1
        fm, body = parse_frontmatter(p)
        if fm is None:
            findings["unparseable_frontmatter"].append({"path": p})
            continue
        is_agent = "/agents/" in p
        desc_len = len(fm.get("_description_raw", ""))
        body_lines = body.count("\n")

        if desc_len > 1024:
            findings["desc_too_long"].append({"path": p, "desc_len": desc_len})

        cap = AGENT_BODY_HARDCAP if is_agent else SKILL_BODY_HARDCAP
        if body_lines > cap:
            findings["body_too_long"].append({"path": p, "body_lines": body_lines, "cap": cap})

        win_drive = re.compile(r"`[^`\n]*[A-Z]:\\[A-Za-z]")
        win_multi = re.compile(r"`[^`\n]*\\[A-Z][A-Za-z0-9_. -]*\\[A-Z]")
        link_drive = re.compile(r"\]\([^)]*[A-Z]:\\[A-Za-z]")
        link_multi = re.compile(r"\]\([^)]*\\[A-Z][A-Za-z0-9_. -]*\\[A-Z]")
        if (win_drive.search(body) or win_multi.search(body)
                or link_drive.search(body) or link_multi.search(body)):
            if p not in WINDOWS_PATH_ALLOWLIST:
                findings["windows_path"].append({"path": p})

        if not is_agent and body_lines > SKILL_BODY_ADVISORY:
            findings["skill_body_over_500"].append({"path": p, "body_lines": body_lines})

    # README component-table vs filesystem: every plugin README lists its
    # components in a markdown table whose rows start `| Skill |` / `| Agent |`.
    # That row count must equal the number of SKILL.md + agent .md files on disk,
    # or the catalog page misleads users browsing the README.
    for readme in sorted(glob.glob("plugins/*/README.md")):
        readme = readme.replace("\\", "/")
        plugin = readme.split("/")[1]
        disk = len(glob.glob(f"plugins/{plugin}/skills/*/SKILL.md")) + len(
            glob.glob(f"plugins/{plugin}/agents/*.md")
        )
        rows = 0
        with open(readme, encoding="utf-8") as fh:
            for line in fh:
                cells = [c.strip() for c in line.split("|")]
                if len(cells) >= 3 and cells[1].lower() in ("skill", "agent"):
                    rows += 1
        if rows != disk:
            findings["readme_count_mismatch"].append(
                {"path": readme, "readme_rows": rows, "disk": disk}
            )

    return findings, total


if __name__ == "__main__":
    import sys
    strict = "--strict" in sys.argv

    findings, total = audit()
    print("# Content audit\n")
    print(f"**Components scanned:** {total}  ")
    print(f"**Mode:** {'strict (fail on critical)' if strict else 'advisory'}\n")

    severity = [
        ("desc_too_long", "CRITICAL"),
        ("body_too_long", "CRITICAL"),
        ("unparseable_frontmatter", "CRITICAL"),
        ("readme_count_mismatch", "WARNING"),
        ("windows_path", "WARNING"),
        ("skill_body_over_500", "WARNING"),
    ]
    critical = 0
    for key, sev in severity:
        items = findings.get(key, [])
        print(f"### [{sev}] {key} — {len(items)}")
        for it in items[:30]:
            print(f"- `{it['path']}`" + "".join(f" ({k}={v})" for k, v in it.items() if k != "path"))
        print()
        if sev == "CRITICAL":
            critical += len(items)

    if strict and critical:
        print(f"[content-audit] {critical} critical finding(s); exiting 1")
        sys.exit(1)
    print(f"[content-audit] {'no critical findings' if not critical else str(critical)+' critical (advisory)'}")

File: scripts/ts/content-audit.ts
#!/usr/bin/env tsx
// Mechanical content checks. Lint guardrails, NOT the quality rubric - the
// six-dimension D1-D6 review is applied by a human reading the diff.
// Port of scripts/content-audit.py; output is byte-identical.

import { readFileSync } from 'node:fs';
import { join } from 'node:path';
import { agents, byCodepoint, exists, pluginNames, skills } from './lib/repo.ts';

const ROOT = process.cwd();
const STRICT = process.argv.includes('--strict');

// Anthropic publishes only "keep SKILL.md under 500 lines"; agents have no
// published cap since they run in their own context. The hard caps below are
// generous ceilings meaning "something is structurally wrong if exceeded".
const SKILL_BODY_HARDCAP = 600;
const AGENT_BODY_HARDCAP = 350;
const SKILL_BODY_ADVISORY = 500;

// Components where a literal Windows path is the vendor-canonical install
// location for a Windows-only tool, not a hygiene oversight. Reviewed 2026-05-27.
const WINDOWS_PATH_ALLOWLIST = new Set([
  'plugins/qa-desktop/skills/appium-windows-driver/SKILL.md',
  'plugins/qa-desktop/skills/winappdriver/SKILL.md',
  'plugins/qa-game/skills/unity-test-framework/SKILL.md',
  'plugins/qa-desktop/skills/desktop-test-strategy-reference/SKILL.md',
]);

type Finding = Record<string, string | number>;

function parseFrontmatter(path: string): [{ desc: string } | null, string] {
  let c: string;
  try {
    c = readFileSync(join(ROOT, path), 'utf8').replace(/\r\n/g, '\n');
  } catch {
    return [null, ''];
  }
  if (!c.startsWith('---')) return [null, c];
  // Python's split("---", 2): parts[1] is the frontmatter, parts[2] the body.
  const i2 = c.indexOf('---', 3);
  if (i2 === -1) return [null, c];
  const fmText = c.slice(3, i2);
  const body = c.slice(i2 + 3);

  const m = /description:\s*"([\s\S]*?)"\s*\n[a-z_-]+:/.exec(fmText);
  const desc = m ? m[1]! : (/^description:\s*"?(.*?)"?\s*$/m.exec(fmText)?.[1] ?? '');
  return [{ desc }, body];
}

const components = [
  ...skills(ROOT).map((s) => s.path),
  ...agents(ROOT).map((a) => a.path),
].sort(byCodepoint);

const findings = new Map<string, Finding[]>();
const add = (key: string, f: Finding): void => {
  const list = findings.get(key) ?? [];
  list.push(f);
  findings.set(key, list);
};

const WIN_DRIVE = /`[^`\n]*[A-Z]:\\[A-Za-z]/;
const WIN_MULTI = /`[^`\n]*\\[A-Z][A-Za-z0-9_. -]*\\[A-Z]/;
const LINK_DRIVE = /\]\([^)]*[A-Z]:\\[A-Za-z]/;
const LINK_MULTI = /\]\([^)]*\\[A-Z][A-Za-z0-9_. -]*\\[A-Z]/;

let total = 0;
for (const p of components) {
  total += 1;
  const [fm, body] = parseFrontmatter(p);
  if (fm === null) {
    add('unparseable_frontmatter', { path: p });
    continue;
  }
  const isAgent = p.includes('/agents/');
  const descLen = fm.desc.length;
  const bodyLines = (body.match(/\n/g) ?? []).length;

  if (descLen > 1024) add('desc_too_long', { path: p, desc_len: descLen });

  const cap = isAgent ? AGENT_BODY_HARDCAP : SKILL_BODY_HARDCAP;
  if (bodyLines > cap) add('body_too_long', { path: p, body_lines: bodyLines, cap });

  if (WIN_DRIVE.test(body) || WIN_MULTI.test(body) || LINK_DRIVE.test(body) || LINK_MULTI.test(body)) {
    if (!WINDOWS_PATH_ALLOWLIST.has(p)) add('windows_path', { path: p });
  }

  if (!isAgent && bodyLines > SKILL_BODY_ADVISORY) {
    add('skill_body_over_500', { path: p, body_lines: bodyLines });
  }
}

// Every plugin README lists its components in a markdown table. That row count
// must equal SKILL.md + agent .md on disk, or the README misleads browsers.
// Rows are counted by the component link they carry rather than the first-cell
// label, so org-chart tables (`| Role | Agent | task |`) count too.
const LINK_RE = /\]\((?:agents\/[^)]+\.md|skills\/[^)]+\/SKILL\.md)\)/;
const allSkills = skills(ROOT);
const allAgents = agents(ROOT);

// Every plugin carrying a README, including role bundles that own no components
// and must therefore show zero component rows.
const pluginsWithReadme = pluginNames(ROOT).filter((p) =>
  exists(join(ROOT, 'plugins', p, 'README.md')),
);

for (const plugin of pluginsWithReadme.sort(byCodepoint)) {
  const readme = `plugins/${plugin}/README.md`;
  const disk =
    allSkills.filter((s) => s.plugin === plugin).length +
    allAgents.filter((a) => a.plugin === plugin).length;
  let rows = 0;
  for (const line of readFileSync(join(ROOT, readme), 'utf8').replace(/\r\n/g, '\n').split('\n')) {
    if (!line.trimStart().startsWith('|')) continue;
    if (line.includes('](../')) continue; // cross-plugin reference row
    const cells = line.split('|').map((c) => c.trim());
    const first = cells.length > 1 ? cells[1]!.toLowerCase() : '';
    if (first === 'skill' || first === 'agent' || LINK_RE.test(line)) rows += 1;
  }
  if (rows !== disk) add('readme_count_mismatch', { path: readme, readme_rows: rows, disk });
}

console.log('# Content audit\n');
console.log(`**Components scanned:** ${total}  `);
console.log(`**Mode:** ${STRICT ? 'strict (fail on critical)' : 'advisory'}\n`);

const SEVERITY: [string, 'CRITICAL' | 'WARNING'][] = [
  ['desc_too_long', 'CRITICAL'],
  ['body_too_long', 'CRITICAL'],
  ['unparseable_frontmatter', 'CRITICAL'],
  ['readme_count_mismatch', 'CRITICAL'],
  ['windows_path', 'WARNING'],
  ['skill_body_over_500', 'WARNING'],
];

let critical = 0;
for (const [key, sev] of SEVERITY) {
  const items = findings.get(key) ?? [];
  console.log(`### [${sev}] ${key} — ${items.length}`);
  for (const it of items.slice(0, 30)) {
    const extra = Object.entries(it)
      .filter(([k]) => k !== 'path')
      .map(([k, v]) => ` (${k}=${v})`)
      .join('');
    console.log(`- \`${it['path']}\`${extra}`);
  }
  console.log('');
  if (sev === 'CRITICAL') critical += items.length;
}

if (STRICT && critical) {
  console.log(`[content-audit] ${critical} critical finding(s); exiting 1`);
  process.exit(1);
}
console.log(`[content-audit] ${critical === 0 ? 'no critical findings' : `${critical} critical (advisory)`}`);
